wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+{
+    char *resp = NULL;
+    int ret;
+    byte len = (byte)sz;
+    if ( !wolfIoT_initialized)
+    {
+        iotsafe_init();
+    }
+    iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_GETRANDOM,0,0);
+    bytes_to_hex(&len, csim_cmd + AT_CSIM_CMD_SIZE + 8, 1);
+    csim_cmd[AT_CSIM_CMD_SIZE + AT_CMD_HDR_SIZE] = 0;
+    iotsafe_cmd_complete(csim_cmd);
+    if (expect_csim_response(csim_cmd, (word32)GETRAND_CMD_SIZE, &resp) < 1) {
+        WOLFSSL_MSG("Unexpected reply from RAND");
+        return WC_HW_E;
+    }
+    ret = hexbuffer_conv(resp, output, sz);
+    if (ret < 0)
+        return ret;
+    expect_tok(NULL, 0, NULL, NULL);
+    return 0; /* WolfSSL API expects 0 */
+}
+
+
+
+#ifdef HAVE_ECC
+
+#define IOTSAFE_ECC_KSIZE 32
+
+static int iotsafe_gen_keypair(byte wr_slot)
+{
+    char *resp;
+    iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_GEN_KEYPAIR, 0, 0);
+    iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_PRIVKEY_ID, 1, &wr_slot);
+    iotsafe_cmd_complete(csim_cmd);
+    if (expect_csim_response(csim_cmd, (word32)strlen(csim_cmd), &resp) < 1) {
+        WOLFSSL_MSG("Unexpected reply from Keygen");
+        return WC_HW_E;
+    }
+    return 0;
+}
+
+static int iotsafe_get_public_key(byte idx, ecc_key *key)
+{
+    int ret;
+    char *resp;
+    char *rkey, *ktype, *payload_str;
+    char Qx[IOTSAFE_ECC_KSIZE * 2 + 1], Qy[IOTSAFE_ECC_KSIZE * 2 + 1];
+
+    /* exporting generated public key */
+    iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_READ_KEY,0, 0);
+    iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_PUBKEY_ID, 1, &idx);
+    iotsafe_cmd_complete(csim_cmd);
+    ret = expect_csim_response(csim_cmd, (word32)strlen(csim_cmd), &resp);
+    if (ret < 1) {
+        WOLFSSL_MSG("Error exporting EPH public key from IoT-Safe");
+        return -1;
+    }
+    if (ret < IOTSAFE_TAG_ECC_KEY_FIELD_SZ + 2) {
+        WOLFSSL_MSG("Response from iot-safe: too short");
+        return -1;
+    }
+    rkey = search_tlv(resp, ret, IOTSAFE_TAG_ECC_KEY_FIELD);
+    if (!rkey) {
+        WOLFSSL_MSG("IoT safe Error in rkey response");
+        return WC_HW_E;
+    }
+    ktype = search_tlv(rkey + 4, IOTSAFE_TAG_ECC_KEY_FIELD_SZ,
+            IOTSAFE_TAG_ECC_KEY_TYPE);
+    if (!ktype) {
+        WOLFSSL_MSG("IoT safe Error in ktype response");
+        return WC_HW_E;
+    }
+    payload_str = search_tlv(ktype + 4, IOTSAFE_TAG_ECC_KEY_TYPE_SZ,
+            IOTSAFE_TAG_ECC_KEY_XY);
+    if (!payload_str) {
+        WOLFSSL_MSG("IoT safe: Error in payload response");
+        return WC_HW_E;
+    }
+    memset(Qx, 0, sizeof(Qx));
+    memset(Qy, 0, sizeof(Qy));
+    ret = wc_ecc_init(key);
+    if (ret < 0) {
+        WOLFSSL_MSG("Cannot initialize ecc key to store IoTSafe public key");
+        return -1;
+    }
+    strncpy(Qx, payload_str + 6, IOTSAFE_ECC_KSIZE * 2);
+    strncpy(Qy, payload_str + 6 + IOTSAFE_ECC_KSIZE * 2, IOTSAFE_ECC_KSIZE * 2);
+
+    if (wc_ecc_import_raw(key, Qx, Qy, NULL, "SECP256R1") < 0) {
```

Comment:
Consider using `wc_ecc_import_raw_ex` with `ECC_SECP256R1`.
`int wc_ecc_import_raw_ex(ecc_key* key, const char* qx, const char* qy, const char* d, int curve_id)`. That avoids the SECP256R1 string lookup. Also consider having a macro in the header to align with `IOTSAFE_ECC_KSIZE` for key type like `#define IOTSAFE_ECC_KTYPE ECC_SECP256R1`.

Reply:
nice one, thanks. Fixed. Also added IOTSAFE_ECC_KTYPE as suggested, in the header.